### ISSUE TITLE
Source-IP is not preserved with externalTrafficPolicy:Local if externalIPs are used

### ISSUE THREAD

**Issue by uablrek** [kubernetes/kubernetes#62537]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/sig network
/area ipvs

**What happened**:

The source address is NATed even when `externalTrafficPolicy: Local` is set in the service when accessed via  `externalIPs`.

**What you expected to happen**:

The source address should be preserved.

**How to reproduce it (as minimally and precisely as possible)**:

- Define a service with `externalIPs` and `externalTrafficPolicy: Local`.
- Connect to the service using a `externalIPs`
- Check the peer address of the connection

**Anything else we need to know?**:

The problem is the `KUBE-MARK-MASQ...match-set KUBE-EXTERNAL-IP` iptables rule;

```
Chain KUBE-SERVICES (2 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 KUBE-MARK-MASQ  all  --  *      *      !11.0.0.0/16          0.0.0.0/0            match-set KUBE-CLUSTER-IP dst,dst
    0     0 KUBE-MARK-MASQ  all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-EXTERNAL-IP dst,dst
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-EXTERNAL-IP dst,dst PHYSDEV match ! --physdev-is-in ADDRTYPE match src-type !LOCAL
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-EXTERNAL-IP dst,dst ADDRTYPE match dst-type LOCAL
```

If that rule is removed with;
```
iptables -t nat -D KUBE-SERVICES 2
```

then the source-ip is preserved.

The source-ip **is** preserved if the service is addresses via the `NodePort`.

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
    Kvm's on Ubuntu 16.04.4 LTS

- OS (e.g. from /etc/os-release):
    own BusyBox

- Kernel (e.g. `uname -a`):
    Linux vm-201 4.15.5 #3 SMP Fri Mar 23 17:46:48 CET 2018 x86_64 GNU/Linux

- Install tools:
    None
- Others:

**Comment by uablrek** [comment]
The traffic is not re-distributed to another node but is passed to a local pod so that part works as expected.

**Comment by uablrek** [comment]
To get `externalIPs` working with `proxy-mode=ipvs` use the work-around proposed in #59976

**Comment by m1093782566** [comment]
@uablrek 

IMO, we have to ALWAYS SNAT packets to external IPs. I don't think it's a bug of IPVS proxier - why not test your case in iptables mode and to see the result?

I think `externalTrafficPolicy: Local` currently only support `LoadBalancer` and `NodePort` services, and does not support `externalIPs`. 

@thockin should we validate the case in API?

**Comment by uablrek** [comment]
I tested, and yes SNAT is used also in `proxy-mode=iptables` even with `externalTrafficPolicy: Local` and service `type: LoadBalancer`.

I have no cloud-provider so nobody is doing something extra with the LoadBalancer services. What this "extra" really is seems to different in GCE/AWS/Azur and of course not for public display. The closest I have got to understand the `type: LoadBalancer` magic is the [metallb](https://metallb.universe.tf/) which implements this function.

But what it does is actually to set the `externalIPs` property in the services with `type: LoadBalancer` and let the kube-proxy do the rest. I can see no difference compared to setting `externalIPs` myself.

Why do you have to "have to ALWAYS SNAT"?

For `externalTrafficPolicy: Local` the packets are never re-routed to another node (that part works) and thus the reply packets will always be handled by the same node. That is really workt without SNAT is also very easy to test, just manually remove the iptables rule.

**Comment by uablrek** [comment]
I tried with the metallb and the iptables rules are different;

```
 pkts bytes target     prot opt in     out     source               destination         
    0     0 KUBE-MARK-MASQ  all  --  *      *      !11.0.0.0/16          0.0.0.0/0            match-set KUBE-CLUSTER-IP dst,dst
    0     0 KUBE-MARK-MASQ  all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set KUBE-LOAD-BALANCER-MASQ dst,dst
    0     0 KUBE-MARK-MASQ  tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp match-set KUBE-NODE-PORT-TCP dst
```

There is more to it when `type: LoadBalancer` is propely implemented than just setting `externalIPs`. This is not a bug. I will close this issue and check the metallb source how it works.

**Comment by micw** [comment]
Hello @uablrek ,

I don't understand why this issue is closed. It's still an issue on recent k8s. It should be a no-brainer to set up an iptables rule that preserves the source IP for a service that is routed to a local pod (using externalTrafficPolicy: Local). Currently this bug breaks documented behaviour (https://kubernetes.io/docs/tutorials/services/source-ip/):

> To avoid this, Kubernetes has a feature to preserve the client source IP (check here for feature availability). Setting service.spec.externalTrafficPolicy to the value Local will only proxy requests to local endpoints, never forwarding traffic to other nodes and thereby preserving the original source IP address

**Comment by uablrek** [comment]
@micw I totally agree. A silly thing is that local policy actually makes the call to be local (this is the hard part) but it still performs a totally unnecessary NAT anyway(!).

When I was working with this our intention was to use "externalIPs" rather than "loadBalanceIP", but we must support src-address preservation. But the response was "that is not supported" for externalIPs (period).

The code change (as you might have noticed) is trivial, but i guess the reluctance to add it is testing. At the time of writing this issue I was not allowed to do PRs myself and loadBalanceIP works, so I closed the issue.

If you plan to make a PR yourself I suggst you include complete e2e tests for support of source-preservation for "externalIPs" to increase the chances to get it accepted.

**Comment by uablrek** [comment]
BTW if you need an easy way to set the loadBalancerIP, please see; https://github.com/Nordix/assign-lb-ip

**Comment by uablrek** [comment]
The passing to the local node for externalIPs and "externalTrafficPolicy: Local" has been corrected somewhere from k8s 1.10. I tested on k8s 1.17.0 and externalIPs are always treated as "Cluster". Sorry for mentioning it without testing if it's still valid.
